src/dataflow/deadlines.rs

```diff
@@ -0,0 +1,123 @@
+use crate::dataflow::{stream::StreamId, time::Timestamp, State};
+use std::{collections::HashMap, sync::Arc};
+use tokio::time::Duration;
+
+/// `CondFn` defines the type of the start and end condition functions.
+/// Each function receives the `ConditionContext` that contains the information necessary for the
+/// evaluation of the start and end conditions (s.a. the message counts and the state of the
+/// watermarks on each stream for the given timestamp)
+type CondFn = dyn Fn(&ConditionContext) -> bool + Send + Sync;
+
+/// Enumerates the types of deadlines available to the user.
+pub enum Deadline {
+    TimestampDeadline(Box<TimestampDeadline>),
+}
+
+#[derive(Debug)]
+pub struct ConditionContext {
+    message_count: HashMap<(StreamId, Timestamp), usize>,
+    watermark_status: HashMap<(StreamId, Timestamp), bool>,
+}
+
+impl ConditionContext {
+    pub fn new() -> Self {
+        ConditionContext {
+            message_count: HashMap::new(),
+            watermark_status: HashMap::new(),
+        }
+    }
+
+    pub fn increment_msg_count(&mut self, stream_id: StreamId, timestamp: Timestamp) {
+        let count = self
+            .message_count
+            .entry((stream_id, timestamp))
+            .or_insert(0);
+        *count += 1;
+    }
+
+    pub fn notify_watermark_arrival(&mut self, stream_id: StreamId, timestamp: Timestamp) {
+        let watermark = self
+            .watermark_status
+            .entry((stream_id, timestamp))
+            .or_insert(false);
+        *watermark = true;
+    }
+
+    pub fn clear_state(&mut self, stream_id: StreamId, timestamp: Timestamp) {
+        self.message_count.remove(&(stream_id, timestamp.clone()));
+        self.watermark_status.remove(&(stream_id, timestamp));
+    }
+}
+
+pub trait DeadlineContextT: Send + Sync {
+    fn get_deadline(&self, ctx: &ConditionContext) -> Duration;
+}
+
+pub trait HandlerContextT: Send + Sync {
```

Comment:
Let's discuss this on call. Essentially my idea was that users may need to maintain some state across handler invocations or may register state from multiple operators and so it makes sense to have a separate `Handler` structure defined that implements the `HandlerContextT`. 

I'm unsure of how good this design is esp. because we don't yet have an idea of how the system will end up managing state in the long term.